### PR TITLE
feat(finetune): add preformatted text dataset support

### LINKED ISSUE #821: [REQ] add preformatted text dataset support
Parent: #214

Depends on #820 and #598.

## Summary

Add a `text` dataset type for preformatted, full-sequence causal-language-model training.

```yaml
datasets:
  - source: organization/domain-corpus
    type: text
```

Expected record:

```json
{"text": "A complete preformatted training sequence."}
```

AIKit must own the sequence-boundary contract so every record has one effective leading BOS where applicable and one terminal EOS.

## Motivation

Preformatted text is common for domain corpora, source code, documents, already-rendered instruction examples, and custom language-modeling sequences.

The frozen Unsloth text path tokenizes `text`, but does not provide the complete public contract AIKit needs:

- EOS is not appended automatically.
- Unsloth 2026.8.1 can duplicate an existing BOS token.
- Tokenizers differ in whether BOS is represented in text or added during tokenization.

This feature is full-sequence text SFT. It must not be described as complete continued pretraining, which needs different LoRA targets and optimizer configuration.

## Proposed changes

- Add a `text` dataset type constant and runtime dispatch.
- Require a non-empty dataset and non-empty string `text` values.
- Add tokenizer-aware normalization:
  - Require a usable EOS token.
  - Append EOS when absent.
  - Do not duplicate terminal EOS.
  - Avoid duplicate effective BOS after tokenization.
  - Do not prepend BOS when the tokenizer adds it automatically.
- Preserve source content other than special-token boundary normalization.
- Continue using `SFTTrainer` with full-sequence loss.
- Ensure packing operates on explicitly EOS-terminated records.
- Add source/row context to validation errors where available.
- Add fixtures, documentation, and a one-step GPU smoke configuration.

## Acceptance criteria

- `type: text` is accepted by Go validation and Python dispatch.
- Text records never use the Alpaca formatter.
- Every normalized record has exactly one terminal EOS.
- Existing EOS is not duplicated.
- Existing or tokenizer-added BOS is not duplicated.
- Tokenizers without BOS remain supported; a tokenizer without EOS fails clearly.
- Null, empty, or non-string values fail with useful errors.
- Full-sequence labels supervise the normalized record.
- Packing retains explicit record boundaries.
- Existing Alpaca behavior remains unchanged.
- A one-step GPU run exports a non-empty GGUF.

## Tests

- Go validation tests for `type: text`.
- Python normalization tests covering missing/existing EOS, explicit/automatic/no BOS, missing EOS token, and invalid values.
- Exact-lock regression proving no duplicate effective BOS and one terminal EOS.
- Add:
  - `test/fixtures/text-smoke.jsonl`
  - `test/aikitfile-unsloth-text-smoke.yaml`
- Run `make test`, `make test-python`, `make lint`, and a GPU fine-tune/export smoke test.

## Out of scope

- Proper continued pretraining.
- Training `embed_tokens` or `lm_head`.
- An embedding-specific learning rate.
- Pretokenized datasets.
- Messages, ShareGPT, or response-only loss.
- Multiple datasets or streaming.
- DPO, GRPO, vision, or audio.

### PR DESCRIPTION
**Pull request by sozercan**
Stacked on #828. This PR targets `issue-598` and should be merged after #828.

## Summary

- add `type: text` as a full-sequence Unsloth SFT dataset format
- normalize tokenizer-aware BOS/EOS boundaries without changing source content beyond boundary tokens
- support tokenizer-added BOS/EOS as well as tokenizers without BOS
- reject records that would lose terminal EOS through truncation using bounded batch tokenization
- redact signed remote dataset details from validation errors
- validate the active Unsloth/TRL preprocessing, packing boundaries, full-sequence labels, and EOS supervision before training
- add Go/Python coverage, docs, a JSONL fixture, and a `text-smoke` fine-tune workflow mode

Existing Alpaca and prompt-completion behavior remains unchanged.

## Tests

- `make test`
- `make test-python` (60 tests)
- `make check-unsloth-lock` with uv 0.12.1
- CI-pinned `golangci-lint` 2.11.2
- `git diff --check`

The GPU `text-smoke` job will be run through the fine-tune workflow on the completed top-of-stack branch.

Closes #821

### CODE REVIEW

File: pkg/finetune/target_unsloth.py [hunk 1]
```diff
@@ -245,27 +259,53 @@ def training_dataset_spec(
...
+    source: str | None = None,
+    record_index: int | None = None,
+) -> str:
+    if source is None:
+        subject = f"{dataset_type} dataset"
+        if record_index is not None:
+            subject += f" record {record_index}"
+        return subject
+
+    subject = f"{dataset_type} dataset source {source!r}"
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Good catch. This was valid on `ff66235`. Follow-up commit `8719e90` replaces raw HTTP(S) sources in AIKit validation subjects with `remote JSON URL`, so projection, schema, empty-record, and `maxSeqLength` errors no longer contain credentials, query strings, signatures, or fragments.

The regression test `test_redacts_signed_url_in_text_validation_error` covers a credential-bearing signed URL, and the complete `train_model` overflow and missing-column paths both emit only the redacted subject. No additional change is needed for this thread. Third-party loader logging is addressed downstream by #833's local materialization and cache hardening.

File: pkg/utils/const.go [hunk 2]
```diff
@@ -14,7 +14,25 @@ const (
 	TargetUnsloth = "unsloth"
 
 	DatasetAlpaca           = "alpaca"
+	DatasetMessages         = "messages"
+	DatasetPreference       = "preference"
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/utils/const.go:18]: This expanded diff is an artifact of the final stacked merge, not an unreviewed scope expansion in the original PR. Each downstream feature remained in its own PR with independent review, unit tests, documentation, and live GPU validation (#831 through #835). During landing, the already-reviewed child branches were merged upward through their feature bases so the stack could be merged into `main`; that made the lower PR's final branch contain the downstream commits. No additional implementation change is needed for this scope-only thread.
